packages/ListBox/src/components/WithTags/useWithTags.js

```diff
@@ -0,0 +1,109 @@
+import React from "react";
+import useI18n from "@paprika/l10n/lib/useI18n";
+import { filter } from "../../helpers/filter";
+
+function prepareDataDictionary(key, data) {
+  const dictionary = {};
+
+  data.forEach(d => {
+    dictionary[d[key]] = d;
+  });
+
+  return dictionary;
+}
+
+export default function useWithTags(
+  key,
+  { defaultData = [], defaultFilteredData = [], defaultSelectedKeys = [], filterAttribute = "label" }
+) {
+  if (!key) throw new Error("Key is Required for useWithTags and should be one of the key (string) in your object");
+
+  const [data, setData] = React.useState(defaultData);
+  const { t } = useI18n();
+  const [selectedKeys, setSelectedKeys] = React.useState(defaultSelectedKeys);
+  const [filteredData, setFilteredData] = React.useState(defaultFilteredData);
+
+  const dataDictionary = React.useMemo(() => {
+    return prepareDataDictionary(key, data);
+  }, [data, key]);
+
+  function handleChange(option, options, selectedOption) {
+    setSelectedKeys(prev => {
+      const prevClone = prev.slice(0);
+      const id = options[selectedOption].content.props[key];
+      if (prev.includes(id)) {
+        prev.splice(prevClone.indexOf(id), 1);
+        return prev;
+      }
+
+      return [...new Set(prevClone.concat([id]))];
+    });
+
+    setFilteredData(defaultData);
+  }
+
+  function getSelectedOptions() {
+    if (selectedKeys.length) {
+      const options = [];
+      selectedKeys.forEach(key => {
+        options.push(dataDictionary[key]);
+      });
+
+      return options;
+    }
+
+    return [];
+  }
+
+  function handleRemove(option) {
+    const index = selectedKeys.indexOf(option[key]);
+    if (index >= 0) {
+      setSelectedKeys(prev => {
+        const clone = prev.slice(0);
+        clone.splice(index, 1);
+        return clone;
+      });
+    }
+  }
+
+  function isSelected(id) {
+    return selectedKeys.includes(id);
+  }
+
+  const handleCustomOption = (func = label => ({ label, isCustom: true })) => label => {
+    // useReducer might be a better alternative
+
+    const option = func(label);
+    setData(prev => {
+      return prev.concat(option);
+    });
+
+    setSelectedKeys(prev => {
+      return prev.concat(option[key]);
+    });
+
+    setData(prevData => {
```

Comment:
Why is this necessary in addition to line 77?

Reply:
```js
    const option = func(label);
    
    setData(prev => {
      return prev.concat(option);
    });

    setSelectedKeys(prev => {
      return prev.concat(option[key]);
    });

    setData(prevData => {
      return prevData.concat(option);
    });
```

Seem I applying two times the same line, brain fart.


But this is the full explanation:
```js
// this create the custom options with the filter's input 
    const option = func(label);
// this add the option to the entire data set and force useMemo function to recreate the dictionary line 26
    setData(prev => {
      return prev.concat(option);
    });
// this add the "key" to the selected item
    setSelectedKeys(prev => {
      return prev.concat(option[key]);
    });

// this restart the options filtered to the original defaultFilterData, the initial state.
    setFilteredData(defaultData);
```